client/setec/store.go

```diff
@@ -0,0 +1,302 @@
+		f map[string]Secret           // :: secret name → fetch function
+	}
+
+	cancel context.CancelFunc // stops the polling task
+	done   <-chan struct{}    // closed when the poller is finished
+
+	// TODO(creachadair): Add metrics.
+}
+
+// StoreConfig is the configuration for Store.
+type StoreConfig struct {
+	// Client is the API client used to fetch secrets from the service.
+	// The service URL must be non-empty.
+	Client Client
+
+	// Secrets are the names of the secrets this Store should retrieve. Only
+	// secrets named here can be read out of the store. An error is reported if
+	// no secrets are listed.
+	Secrets []string
+
+	// Cache, if non-nil, is a cache that persists secrets locally.
+	//
+	// Depending on the implementation, local caching may degrade security
+	// slightly by making secrets easier to get at, but in return allows the
+	// Store to initialize and run during outages of the secrets management
+	// service.
+	//
+	// If no cache is provided, the Store caches secrets in-memory for the
+	// lifetime of the process only.
+	Cache Cache
+
+	// PollInterval is the interval at which the store will poll the service for
+	// updated secret values. If zero or negative, a default value is used.
+	PollInterval time.Duration
+
+	// Logf is a logging function where text logs should be sent.  If nil, logs
+	// are written to the standard log package.
+	Logf logger.Logf
+}
+
+func (c StoreConfig) logger() logger.Logf {
+	if c.Logf == nil {
+		return log.Printf
+	}
+	return c.Logf
+}
+
+func (c StoreConfig) pollInterval() time.Duration {
+	if c.PollInterval <= 0 {
+		return 1 * time.Minute
+	}
+	return c.PollInterval
+}
+
+func (c StoreConfig) cache() Cache { return c.Cache }
+
+// NewStore creates a secret store with the given configuration.  The service
+// URL of the client must be set.
+//
+// NewStore blocks until all the secrets named in cfg.Secrets are available for
+// retrieval by the Secret method, or ctx ends.  The context passed to NewStore
+// is only used for initializing the store. If a cache is provided, cached
+// values are accepted even if stale, as long as there is a value for each of
+// the secrets in cfg.
+func NewStore(ctx context.Context, cfg StoreConfig) (*Store, error) {
+	if cfg.Client.Server == "" {
+		return nil, errors.New("no service URL is set")
+	} else if len(cfg.Secrets) == 0 {
+		return nil, errors.New("no secrets are listed")
+	}
+
+	s := &Store{
+		client: cfg.Client,
+		logf:   cfg.logger(),
+		cache:  cfg.cache(),
+	}
+
+	// Initialize the active versions maps.
+	s.active.m = make(map[string]*api.SecretValue)
+	s.active.f = make(map[string]Secret)
+
+	// If we have a cache, try to load data from there first.
+	data, err := s.cache.Read()
+	if err != nil {
+		return nil, fmt.Errorf("reading cached secrets: %w", err)
+	}
+	if len(data) != 0 { // it's OK for the cache to be empty
+		if err := json.Unmarshal(data, &s.active.m); err != nil {
+			return nil, fmt.Errorf("invalid cache: %w", err)
```

Comment:
Similar to above, should an invalid cache degrade to online initialization + healing the cache, rather than a crashloop?